plugins/docker/docker_stats_input.go

```diff
@@ -0,0 +1,216 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# The Initial Developer of the Original Code is the Mozilla Foundation.
+# Portions created by the Initial Developer are Copyright (C) 2014
+# the Initial Developer. All Rights Reserved.
+#
+# Contributor(s):
+#   Anton Lindström (carlantonlindstrom@gmail.com)
+#   Rob Miller (rmiller@mozilla.com)
+#   Karl Matthias (karl.matthias@gonitro.com)
+#   Guy Templeton (guy.templeton@skyscanner.net)
+#
+# ***** END LICENSE BLOCK *****/
+
+package docker
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"math"
+	"github.com/mozilla-services/heka/pipeline"
+	"github.com/mozilla-services/heka/message"
+	"github.com/pborman/uuid"
+)
+
+type DockerStatsInputConfig struct {
+	// A Docker endpoint.
+	Endpoint      		string   `toml:"endpoint"`
+	CertPath      		string   `toml:"cert_path"`
+	NameFromEnv   		string   `toml:"name_from_env_var"`
+	FieldsFromEnv 		[]string `toml:"fields_from_env"`
+	FieldsFromLabels 	[]string `toml:"fields_from_labels"`
+}
+
+type DockerStatsInput struct {
+	conf         *DockerStatsInputConfig
+	stopChan     chan error
+	closer       chan struct{}
+	statsstream  chan *DockerStat
+	attachErrors chan error
+	statsMgr     *StatsManager
+	pack	     *pipeline.PipelinePack
+}
+
+func (di *DockerStatsInput) ConfigStruct() interface{} {
+	return &DockerStatsInputConfig{
+		Endpoint: "unix:///var/run/docker.sock",
+		CertPath: "",
+	}
+}
+
+func (di *DockerStatsInput) Init(config interface{}) error {
+
+	di.conf = config.(*DockerStatsInputConfig)
+	di.stopChan = make(chan error)
+	di.closer = make(chan struct{})
+	di.statsstream = make(chan *DockerStat)
+	di.attachErrors = make(chan error)
+
+	m, err := NewStatsManager(di.conf.Endpoint, di.conf.CertPath, di.attachErrors, di.conf.NameFromEnv, di.conf.FieldsFromEnv, di.conf.FieldsFromLabels)
+	if err != nil {
+		return fmt.Errorf("DockerStatsInput: failed to attach: %s", err.Error())
+	}
+
+	di.statsMgr = m
+	return nil
+}
+
+func (di *DockerStatsInput) Run(ir pipeline.InputRunner, h pipeline.PluginHelper) error {
+	var ok   bool
+	di.statsMgr.ir = ir
+	hostname := h.Hostname()
+
+	go di.statsMgr.Run(di.statsstream, di.closer, di.stopChan)
+
+	err := withRetries(func() error { return di.statsMgr.client.AddEventListener(di.statsMgr.events) })
+	// Get the InputRunner's chan to receive empty PipelinePacks
+	packSupply := ir.InChan()
+
+	ok = true
+	for ok {
+		select {
+		case statsline := <-di.statsstream:
+			di.pack = <-packSupply
+
+			di.pack.Message.SetType("DockerStats")
+			di.pack.Message.SetLogger(statsline.Container)
+			di.pack.Message.SetHostname(hostname)   // Use the host's hosntame
+			di.pack.Message.SetPayload(statsline.StatsString)
+			di.pack.Message.SetTimestamp(statsline.Time.UnixNano())
+			di.pack.Message.SetUuid(uuid.NewRandom())
+			di.MarshalStatsFields("stat", reflect.ValueOf(statsline.Stat), ir)
```

Comment:
Sorry for showing up late here, and having feedback that undoes some of the work that you've done. However, I don't think we should be writing any of the stats values to fields here in the input plugin. That work would be better done in a SandboxDecoder. You can parse the JSON and extract the values from there to add to the message you're injecting.